src/main/kotlin/com/openlattice/authorization/mapstores/ResolvedPrincipalTreesMapLoader.kt

```diff
@@ -0,0 +1,101 @@
+import com.openlattice.organizations.roles.SecurePrincipalsManager
+import org.apache.commons.lang3.NotImplementedException
+import org.apache.commons.lang3.RandomStringUtils
+import java.util.*
+import javax.inject.Inject
+
+/**
+ * In memory cache of resolved principal trees
+ * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
+ */
+class ResolvedPrincipalTreesMapLoader : TestableSelfRegisteringMapStore<String, NavigableSet<Principal>> {
+
+    @Inject
+    private lateinit var spm: SecurePrincipalsManager
+    private lateinit var principals: IMap<String, SecurablePrincipal>
+
+    override fun getMapName(): String {
+        return HazelcastMap.RESOLVED_PRINCIPAL_TREES.name
+    }
+
+    override fun loadAllKeys(): Iterable<String>? {
+        return null
+    }
+
+    override fun getTable(): String {
+        return ""
+    }
+
+    override fun generateTestValue(): NavigableSet<Principal> {
+        return ImmutableSortedSet.of(TestDataFactory.userPrincipal())
+    }
+
+    override fun store(key: String, value: NavigableSet<Principal>) {
+        throw NotImplementedException("This is a read only map loader.")
+    }
+
+    override fun generateTestKey(): String {
+        return RandomStringUtils.random(10)
+    }
+
+    override fun deleteAll(keys: Collection<String>) {
+        throw NotImplementedException("This is a read only map loader.")
+    }
+
+    override fun storeAll(map: Map<String, NavigableSet<Principal>>) {
+        throw NotImplementedException("This is a read only map loader.")
+    }
+
+    override fun getMapConfig(): MapConfig {
+        return MapConfig(mapName)
+                .setMaxIdleSeconds(300)
+                .setMapStoreConfig(mapStoreConfig)
+    }
+
+    override fun getMapStoreConfig(): MapStoreConfig {
+        return MapStoreConfig()
+                .setEnabled(true)
+                .setImplementation(this)
+                .setInitialLoadMode(MapStoreConfig.InitialLoadMode.EAGER)
+    }
+
+    override fun loadAll(keys: MutableCollection<String>): Map<String, NavigableSet<Principal>> {
+        return keys.associateWith { load(it) }
+    }
+
+    override fun load(principalId: String): NavigableSet<Principal> {
+        val sp = principals[principalId] ?: return ImmutableSortedSet.of()
+        val securablePrincipals = spm.getAllPrincipals(sp) ?: return ImmutableSortedSet.of()
+
+        val currentPrincipals: NavigableSet<Principal> = TreeSet()
+        currentPrincipals.add(sp.principal)
+        securablePrincipals.stream()
+                .map(SecurablePrincipal::getPrincipal)
+                .forEach { currentPrincipals.add(it) }
+        return currentPrincipals
+    }
+
+    override fun delete(key: String) {
+        throw NotImplementedException("This is a read only map loader.")
+    }
+
+    @Inject
+    fun initPrincipalsMapstore(hazelcastInstance: HazelcastInstance) {
+        this.principals = hazelcastInstance.getMap(HazelcastMap.SECURABLE_PRINCIPALS.name)
```

Comment:
I'm a little confused as to why we have this whole class that wraps an IMap. Is it just that there's no way to configure hazelcast to configure `HazelcastMap.SECURABLE_PRINCIPALS` as a `NearCache`?

Reply:
I think this is out of date now. This class specifically also does more and uses SecurablePrincipalManager to walk tree of all children into a sorted set of SecurablePrincipals.

Reply:
i.e it leverages multiple underlying maps and caches the computation (PrincipalTrees and Principals) based on id lookup.